### PR TITLE
Implement masonry-auto-flow for CSS Grid Masonry Layout.

### PR DESCRIPTION
**Pull request by sammygill**
#### 1e74b53c5b2ba2cf8695b05277587845053adc42
<pre>
Implement masonry-auto-flow for CSS Grid Masonry Layout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247079">https://bugs.webkit.org/show_bug.cgi?id=247079</a>
rdar://101597425

Reviewed by Brent Fulgham.

The masory-auto-flow property modifies how items are placed during
masonry layout. If the property specifies &quot;next,&quot; then items will be
placed in the next available position rather than finding the grid axis
track with most available space. If the property specifies &quot;ordered,&quot;
then all items after first track items are placed using order-modified
document order rather prioritizing items with definite grid axis
positions.

In order to achieve the &quot;next,&quot; functionality, all we need to do is
keep track of a cursor that points to the next available track to
place an item. This cursor will be reset to 0 after the first track
items have been placed. After that, the cursor will be updated after
each item has placed. When an item that does not have a definite grid
axis position needs to be placed, we will then reference the cursor.

In order to achieve the &quot;ordered,&quot; functionality, we modify the code to
take a different patch after the first track items have been placed. In
this new path we will use the grid iterator to get the correct order
of items. For each item we will need to check if the item has a definite
grid axis position or not to position the item correctly. In this
scenario we can also avoid allocating memory for the grid vectors that
hold the definite and indefinite grid axis positions items since we will
not be needing them. They were being used to help prioritize the two
different types of items during placement, but in this scenario we do
not need to do that.

Spec reference: <a href="https://drafts.csswg.org/css-grid-3/#masonry-auto-flow">https://drafts.csswg.org/css-grid-3/#masonry-auto-flow</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-columns-item-placement-auto-flow-next-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-columns-item-placement-auto-flow-next-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-rows-item-placement-auto-flow-next-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-rows-item-placement-auto-flow-next-001.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::performMasonryPlacement):
(WebCore::GridMasonryLayout::collectMasonryItems):
(WebCore::GridMasonryLayout::allocateCapacityForMasonryVectors):
(WebCore::GridMasonryLayout::placeItemsUsingOrderModifiedDocumentOrder):
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::positionForDefiniteGridAxisItem const): Deleted
(WebCore::GridMasonryLayout::gridAreaForDefiniteGridAxisItem const):
(WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
(WebCore::GridMasonryLayout::updateRunningPositions):
Rename varaible for a little more clarity that the span is in the grid
axis direction.

(WebCore::GridMasonryLayout::gridAxisPositionUsingPackAutoFlow const):
(WebCore::GridMasonryLayout::gridAxisPositionUsingNextAutoFlow):
(WebCore::GridMasonryLayout::positionForIndefiniteGridAxisItem): Deleted
(WebCore::GridMasonryLayout::gridAreaForIndefiniteGridAxisItem):
Depending on the value of masonry-auto-flow, we will need to see which
code path to use to determine the position of items that have an
indefinite position in the grid axis.

(WebCore::GridMasonryLayout::gridAxisSpanFromArea const):
(WebCore::GridMasonryLayout::masonryGridAreaFromGridAxisSpan const):
(WebCore::GridMasonryLayout::hasEnoughSpaceAtPosition const):
(WebCore::GridMasonryLayout::nextMasonryPositionForItem): Deleted.
* Source/WebCore/rendering/GridMasonryLayout.h:

Canonical link: <a href="https://commits.webkit.org/257643@main">https://commits.webkit.org/257643@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/51be1feb1400549882d38f86fce0e4d46410f11b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99549 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8724 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108911 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169148 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86029 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92024 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106828 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105304 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34006 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21909 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2566 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23424 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2497 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8648 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42899 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5260 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4357 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by sammygill** [comment]
Parsing patch is approved, just waiting for EWS to land in https://github.com/WebKit/WebKit/pull/6978 so we can ignore those bits in this review

File: Source/WebCore/css/ComputedStyleExtractor.cpp
Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp:3275]: Do we need to append for the pack case?

File: Source/WebCore/css/ComputedStyleExtractor.cpp
Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Is definite-first supposed to mean packed? The spec was kind of unclear on this when I read it.

File: Source/WebCore/css/ComputedStyleExtractor.cpp
Comment (Loirooriol) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Maybe store `style.masonryAutoFlow()` in a variable with a shorter name?

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (Loirooriol) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: To avoid repeating the condition, this could be

```cpp
else if (m_renderGrid.style().masonryAutoFlow().contains(MasonryAutoFlow::DefiniteFirst)) {
    if (hasDefiniteGridAxisPosition(child, gridAxisDirection()))
        m_itemsWithDefiniteGridAxisPosition.append(child);
    else
        m_itemsWithIndefiniteGridAxisPosition.append(child);
}
```

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (Loirooriol) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: It's a bit strange that in this case you have to call `masonryGridAreaFromGridAxisSpan`, while in the other case below it's already called by `positionForIndefiniteGridAxisItem`.

Why not rename `gridAxisSpanForItemWithDefiniteGridAxisPosition` to `positionForDefiniteGridAxisItem` and let it call `masonryGridAreaFromGridAxisSpan`?

File: Source/WebCore/rendering/style/RenderStyleConstants.h
Comment (Loirooriol) [inline Source/WebCore/rendering/style/RenderStyleConstants.h]: Unneeded empty line

**Review by Loirooriol — COMMENTED** [review]
Just some minor suggestions, I don't have time now to look at the details.

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (mattwoodrow) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: Is it potentially a performance problem that we have to do a hash map lookup into the first track items, for every child?

Could instead `collectMasonryItems` just put all non-first track items into one of the two definite/indefinite lists (doesn't matter which, just unconditionally put them all in one if definite-first is enabled).

We could then iterate that list in this function, knowing that it already excludes the first track children (and is in the right order).

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (mattwoodrow) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: What happens in the case where the item has a definite span that is larger than the explicit grid, and m_autoFlowNextCursor is not 0.

For example, say we have (in the grid axis), 3 explicit columns, and an item that has requested 'span 5'.

The initial 'ignoring masonry' placement step should have run Step 3 of the [Grid Item placement algorithm](https://drafts.csswg.org/css-grid-2/#auto-placement-algo), which will determine that we need 2 extra implicit columns for that item to fit. I'd expect m_gridAxisTracksCount to be 5 as a result of that.

If we're doing next placement, and m_autoFlowNextCursor is not-0, then this code will result in the item extending past the end of the implicit grid.
Inserting in the grid modifies the implicit grid, but it won't change m_gridAxisTracksCount, and they'll be out of sync.

I can't tell what the spec wants here. For non-masonry placement, we'd reset back to 0 if there wasn't enough room on the line to fit the item, rather than extending the implicit grid a second time.

Comment (mattwoodrow) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: From section 2.3: " plus the items that span all grid axis tracks (which obviously will be placed at the first implicit line since there is no other way to place them). "

File: Source/WebCore/animation/CSSPropertyAnimation.cpp
Comment (graouts) [inline Source/WebCore/animation/CSSPropertyAnimation.cpp]: Please file a bug to add discrete animation support for `masonry-auto-flow`.

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Done

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Fixed

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Yeah I could see that becoming an issue. Changed the logic to store all of the items in the definite items vector.

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Yeah I think it makes sense for the cursor to reset back to the beginning of the grid if the item would span out of the grid. I added a couple of test cases to cover this scenario.

File: Source/WebCore/animation/CSSPropertyAnimation.cpp
Comment (sammygill) [reply Source/WebCore/animation/CSSPropertyAnimation.cpp]: Done https://bugs.webkit.org/show_bug.cgi?id=248646

File: LayoutTests/TestExpectations
Comment (brentfulgham) [inline LayoutTests/TestExpectations]: Is all of this just a whitespace change? Maybe this doesn't need to be in the PR?

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (brentfulgham) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: Nit: The name of this method makes me think it returns a position (e.g., [x,y] location) of some kind, but it returns an area (e.g, [width, height], or possibly [x, y, width, height]). Perhaps this is the nomenclature used throughout Grid code, but I was slightly confused while reading.

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (brentfulgham) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: Nit: You dereference 'item' without a nullptr check; maybe this should just take a `const RenderBox&` argument, since it seems unexpected this method would ever get a nullptr?

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (brentfulgham) [inline Source/WebCore/rendering/GridMasonryLayout.cpp]: Maybe this should take a `const RenderBox&` argument, since you just dereference items without a nullptr check.

**Review by brentfulgham — APPROVED** [review]
LGTM, but I think a grid expert should review, too.

File: LayoutTests/TestExpectations
Comment (sammygill) [reply LayoutTests/TestExpectations]: Whitespace and a new test that we are failing. I created 2 new tests, but 1 of them we are failing because of an issue that is unrelated to the part of the spec it is testing. I actually think it will get fixed with another patch I'm finishing up. I'll fix the whitespace in another patch

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: I don't think position here refers to a specific grid concept so this could be changed to something like `gridAreaForDefiniteGridAxisItem`

CC @mattwoodrow in case I am wrong on this though

File: Source/WebCore/css/ComputedStyleExtractor.cpp
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: The spec if a little weird with this. `definite-first` is the default state for the order unless `ordered` is specified. So if you do not specify `ordered,` the behavior is always `definite-first` regardless of whether it is `pack` or `next`

Comment (stwrt) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Can we add a comment then 😀?

Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Done!

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Fixed by https://github.com/WebKit/WebKit/pull/7357

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Fixed by https://github.com/WebKit/WebKit/pull/7357

**Comment by sammygill** [comment]
Talked to Matt and he thinks it looks good to land. Did a rebase so just going to wait for EWS as a sanity check.

File: Source/WebCore/rendering/GridMasonryLayout.cpp
Comment (sammygill) [reply Source/WebCore/rendering/GridMasonryLayout.cpp]: Ended up changing the names because I agree it probably it more intuitive

**Pull request merged**
